datalad/support/gitrepo.py

```diff
@@ -416,6 +418,213 @@ def wrapped(repo, *args, **kwargs):
     return wrapped
 
 
+class GitProgress(object):
+    """Reduced variant of GitPython's RemoteProgress class
+
+    Original copyright:
+        Copyright (C) 2008, 2009 Michael Trier and contributors
+    Original license:
+        BSD 3-Clause "New" or "Revised" License
+    """
+    _num_op_codes = 10
+    BEGIN, END, COUNTING, COMPRESSING, WRITING, RECEIVING, RESOLVING, FINDING_SOURCES, CHECKING_OUT, ENUMERATING = \
+        [1 << x for x in range(_num_op_codes)]
+    STAGE_MASK = BEGIN | END
+    OP_MASK = ~STAGE_MASK
+
+    DONE_TOKEN = 'done.'
+    TOKEN_SEPARATOR = ', '
+
+    _known_ops = {
+        COUNTING: ("Counting", "Objects"),
+        ENUMERATING: ("Enumerating", "Objects"),
+        COMPRESSING: ("Compressing", "Objects"),
+        WRITING: ("Writing", "Objects"),
+        RECEIVING: ("Receiving", "Objects"),
+        RESOLVING: ("Resolving", "Deltas"),
+        FINDING_SOURCES: ("Finding", "Sources"),
+        CHECKING_OUT: ("Check out", "Things"),
+    }
+
+    __slots__ = ('_seen_ops', '_pbars')
+
+    re_op_absolute = re.compile(r"(remote: )?([\w\s]+):\s+()(\d+)()(.*)")
+    re_op_relative = re.compile(r"(remote: )?([\w\s]+):\s+(\d+)% \((\d+)/(\d+)\)(.*)")
+
+    def __init__(self):
+        self.__enter__()
+
+    def __enter__(self):
+        self._seen_ops = []
+        self._pbars = set()
+        return self
+
+    def __exit__(self, exc_type, exc_value, traceback):
+        # take down any progress bars that were not closed orderly
+        for pbar_id in self._pbars:
+            log_progress(
+                lgr.info,
+                pbar_id,
+                'Finished',
+            )
+
+    def __call__(self, byts):
+        """Callable interface compatible with WitlessRunner()
+
+        Parameters
+        ----------
+        byts : bytes
+          One or more lines of command output.
+
+        Returns
+        -------
+        bytes
+          All input in its original form, except for lines that
+          were identified as recognized progress reports.
+        """
+        keep_lines = []
+        for line in byts.splitlines(keepends=True):
+            if not self._parse_progress_line(line):
```

Comment:
Hmm, so we're potentially dealing with a partial line here.  For example, on one run of clone I see a line `r` (which is returned as non-progress output) followed by a line "emote: Counting objects:  86% [...]" (also returned as non-progress output).  If we're just losing a counting line here and there, that isn't too concerning, I suppose.  Each segment of a partial error/warning/fatal line is unlikely to have a progress match, so both partial lines should end up in `keep_lines`.  `keep_lines` will be joined back together, so downstream code will still be able to detect, say, a "warning: " line.

So I'm not sure how much of a problem this is.  But I think it at least deserves a comment.

Reply:
I added a comment.